### PR TITLE
Remove fixme in nodeAgg.c:ExecSquelchAgg.

### PR DESCRIPTION
**Pull request by z-wenlin**
As the comment mentioned, for the test query from 'eagerfree' test, it will break before.
```
with my_group_sum(d, total) as (select d, sum(i) from smallt group by d)
select smallt2.* from smallt2
where 0 < all (select total from my_group_sum, smallt2 as tmp where my_group_sum.d = smallt2.d);
```
The query plan looks like
```
                                                      QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1.50 rows=25 width=15)
   ->  Seq Scan on smallt2  (cost=0.00..1.17 rows=8 width=15)
         Filter: (SubPlan 1)
         SubPlan 1
           ->  Nested Loop  (cost=10000000003.67..10000000019.17 rows=1050 width=8)
                 ->  Result  (cost=3.67..4.09 rows=21 width=12)
                       Filter: (smallt.d = smallt2.d)
                       ->  HashAggregate  (cost=3.67..3.88 rows=21 width=12)
                             Group Key: smallt.d
                             ->  Materialize  (cost=0.00..3.17 rows=100 width=8)
                                   ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..2.67 rows=100 width=8)
                                         ->  Seq Scan on smallt  (cost=0.00..1.33 rows=33 width=8)
                 ->  Materialize  (cost=0.00..2.08 rows=50 width=0)
                       ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=0.00..1.83 rows=50 width=0)
                             ->  Seq Scan on smallt2 tmp  (cost=0.00..1.17 rows=17 width=0)
 Optimizer: Postgres query optimizer
(16 rows)
```
Sometimes, when `ExecSquelchAgg()` is called, if `ExecEagerFreeAgg()` here, it will destroy the hash context. When the node is rescanning, it will call `ExecReScanAgg()`,`resetTupleHashIterator()` will raise error.
```
 /*
  * If we do have the hash table, and it never spilled, and the subplan
  * does not have any parameter changes, and none of our own parameter
  * changes affect input expressions of the aggregated functions, then
  * we can just rescan the existing hash table; no need to build it
  * again.
  */
  if (outerPlan->chgParam == NULL && !node->hash_ever_spilled &&
      !bms_overlap(node->ss.ps.chgParam, aggnode->aggParams))
  {
      ResetTupleHashIterator(node->perhash[0].hashtable,
                          &node->perhash[0].hashiter);
      select_current_set(node, 0, true);
      return;
  }
```
```
#2  0x0000000000a5b329 in CdbProgramErrorHandler (postgres_signal_arg=11) at postgres.c:3640
#3  <signal handler called>
#4  0x00000000007b6c86 in tuplehash_start_iterate (tb=0x7f7f7f7f7f7f7f7f, iter=0x2fa56d8) at ../../../src/include/lib/simplehash.h:839
#5  0x00000000007e0c98 in ExecReScanAgg (node=0x2f6b418) at nodeAgg.c:4679
#6  0x00000000007a4ecb in ExecReScan (node=0x2f6b418) at execAmi.c:294
```

The reason why GP6 doesn't have this problem is that it doesn't reuse the hash table, and this reuse hash table logic brought after merging PG12 is a kind of optimization.

In order to keep ExecEagerFreeAgg() optimization inside ExecSquelchAgg(), but also fix this kind of broken query, this PR checks if subplan has any params change, if no change, it will not clean the hash and reuse the hash table later.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Review by adam8157 — CHANGES_REQUESTED** [review]
-1

`ExecEagerFreeAgg()` is an important optimization, please don't remove it easily.

Are we sure that Squelch is being called rightly?

File: src/backend/executor/nodeAgg.c [hunk 1]
```diff
@@ -5021,26 +5021,5 @@ ExecEagerFreeAgg(AggState *node)
...
-	 * That was broken, because ExecEagerFreeAgg() currently doesn't set
-	 * table_filled=false, even though it destroys the 'hashcontext'.
-	 * But it also doesn't destroy the hash table itself (or does go away
-	 * along with 'hashcontext'?) Freeing stuff aggressively here seems like
-	 * a premature optimization to me, so instead of trying to fix all that
-	 * I just disabled this attempt at eagerly freeing stuff. Revisit later?
-	 * I've got a feeling that Squelch is being called when it shouldn't, in
-	 * queries like above that involve a SubPlan.
-	 */
-	//ExecEagerFreeAgg(node);
```
Comment (zxuejing) [inline src/backend/executor/nodeAgg.c:5044]: I see in ExecRescanAgg that the below situation will use hashtable again.
```
/*
 * If we do have the hash table, and it never spilled, and the subplan
 * does not have any parameter changes, and none of our own parameter
 * changes affect input expressions of the aggregated functions, then
 * we can just rescan the existing hash table; no need to build it
 * again.
 */
if (outerPlan->chgParam == NULL && !node->hash_ever_spilled &&
	!bms_overlap(node->ss.ps.chgParam, aggnode->aggParams))
{
	ResetTupleHashIterator(node->perhash[0].hashtable,
					&node->perhash[0].hashiter);
	    select_current_set(node, 0, true);
		return;
}
```
so can we in ExecSquelchAgg add the below codes? 
we just call ExecEagerFreeAgg here when we do not use hashtable again.
```
PlanState  *outerPlan = outerPlanState(node);
Agg     *aggnode = (Agg *) node->ss.ps.plan;
if (!(outerPlan->chgParam == NULL && !node->hash_ever_spilled &&
   !bms_overlap(node->ss.ps.chgParam, aggnode->aggParams)))
      ExecEagerFreeAgg(node);
```

Comment (adam8157) [reply src/backend/executor/nodeAgg.c:5044]: I'd like to confirm if we call `ExecSquelchAgg()` wrongly first.

The good behavior is not to call `ExecSquelchAgg()` if not supposed to, else call `ExecSquelchAgg()` and `ExecEagerFreeAgg()` without check, I think.

**Comment by Aegeaner** [comment]
I have read the following comments before `void ExecSquelchNode(PlanState *node);` and thought it helpful:

```
 * After a node has been squelched, you mustn't try to read more tuples from
 * it. However, ReScanning the node will "un-squelch" it, allowing to read
 * again. Squelching a node is roughly equivalent to fetching and discarding
 * all tuples from it.
```
https://github.com/greenplum-db/gpdb/blob/806fc86bc12d181b3f870da9e393296bc4e01030/src/backend/executor/execAmi.c#L674

**Comment by z-wenlin** [comment]
Thanks all for the review and comments.
As for the Squelch, when the outerTupleSlot is null, it will call `ExecSquelchNode() -> ExecSquelchAgg()`,
```
ExecNestLoop(PlanState *pstate)
{  
    if (TupIsNull(result))
    {
        ExecSquelchNode(pstate);
    }    
}
```
If we call `ExecEagerFreeAgg()` inside `ExecSquelchAgg()`, it will clean hash context.
```
    if (node->hash_metacxt != NULL)
    {
        MemoryContextDelete(node->hash_metacxt);
        node->hash_metacxt = NULL;
    }
```
After that, `ExecNestLoop_guts() -> ExecProcNode(outerPlan)` (Result node has chParams, for example smallt2.d)
Then `ExecReScan() -> ExecReScanResult() -> ExecReScanAgg()`
When execute `ExecReScanAgg()`, it will not rebuild the hash and try to resetTupleHashIterator, then it will panic for this kind of query.
```
 /*
  * If we do have the hash table, and it never spilled, and the subplan
  * does not have any parameter changes, and none of our own parameter
  * changes affect input expressions of the aggregated functions, then
  * we can just rescan the existing hash table; no need to build it
  * again.
  */
  if (outerPlan->chgParam == NULL && !node->hash_ever_spilled &&
      !bms_overlap(node->ss.ps.chgParam, aggnode->aggParams))
  {
      ResetTupleHashIterator(node->perhash[0].hashtable,
                          &node->perhash[0].hashiter);
      select_current_set(node, 0, true);
      return;
  }

```

After comparing it with GP6, the reason why 6 doesn't have this problem is that it doesn't reuse the hash table, which means if it's hashed, it will destroy the hash table and rebuild the hash during the rescan, and this part of reuse hash logic is new on master, which is brought after the upstream merge, and I think this feature is an enhancement.

As for whether we should keep the ExecEagerFreeAgg() for optimization, I think we could try some alternative solutions, such as the solutions like xuejing mentioned above, but still need some tests.

**Comment by adam8157** [comment]
Great analysis, now @zxuejing 's proposal looks awesome.

Please go through each check to confirm.

**Review by adam8157 — APPROVED** [review]

File: src/backend/executor/nodeAgg.c [hunk 2]
```diff
@@ -5022,25 +5021,26 @@ void
...
+
+	if (!ReuseHashTable(node))
+	{
+		ExecEagerFreeAgg(node);
+	}
+
 	ExecSquelchNode(outerPlanState(node));
 }
+
+bool
```
[the remark this answers is not in the document]
Comment (adam8157) [reply src/backend/executor/nodeAgg.c:5039]: oops, missed this, sorry

**Pull request merged**